hubblestack/extmods/modules/audit.py

```diff
@@ -1,450 +1,702 @@
+        log.debug('Executing check-id: %s in nova profile: %s', audit_id, nova_profile)
+        audit_impl = _get_matched_implementation(audit_id, audit_data, tags, labels)
+        if not audit_impl:
+            # no matched impl found
+            continue
 
+        if not _validate_audit_data(audit_id, audit_impl):
+            continue
 
-def _load_audit_file(audit_file, path):
-    """ Load the current audit file"""
-    audit_data = None
-    if os.path.isfile(path):
         try:
-            with open(path, 'r') as audit_file_handler:
-                audit_data = yaml.safe_load(audit_file_handler)
+            # version check
+            if not _is_audit_check_version_compatible(audit_id, audit_impl):
+                raise AuditCheckVersionIncompatibleError('Version not compatible')
+
+            if _is_boolean_expression(audit_impl):
+                # Check is boolean expression.
+                # Gather boolean expressions in separate list and evaluate after evaluating all other checks.
+                log.debug('Boolean expression found. Gathering it to evaluate later.')
+                boolean_expr_dict ={}
+                boolean_expr_dict['audit_id'] = audit_id
+                boolean_expr_dict['audit_impl'] = audit_impl
+                boolean_expr_dict['audit_data'] = audit_data
+                boolean_expr_check_list.append(boolean_expr_dict)
+            else:
+                # handover to module
+                audit_result = _execute_module(audit_id, audit_impl, audit_data, verbose, nova_profile)
+                result_list.append(audit_result)
+        except AuditCheckValidationError as validation_error:
+            # add into error section
+            error_dict={}
+            error_dict['tag'] = audit_data['tag']
+            error_dict['description'] = audit_data['description']
+            error_dict['check_result'] = CHECK_STATUS['Error']
+            error_dict['nova_profile'] = nova_profile
+            result_list.append(error_dict)
+            log.error(validation_error)
+        except AuditCheckVersionIncompatibleError as version_error:
+            # add into skipped section
+            skipped_dict = {}
+            skipped_dict['tag'] = audit_data['tag']
+            skipped_dict['description'] = audit_data['description']
+            skipped_dict['check_result'] = CHECK_STATUS['Skipped']
+            skipped_dict['nova_profile'] = nova_profile
+            result_list.append(skipped_dict)
+            log.error(version_error)
         except Exception as exc:
-            log.exception('Error loading audit file %s: %s', audit_file, exc)
-            return False, audit_data
-    if not audit_data or not isinstance(audit_data, dict):
-        log.error('audit data from %s was not formed as a dict', audit_file)
-        return False, audit_data
-    return True, audit_data
+            log.error(exc)
+
+    #Evaluate boolean expressions
+    if boolean_expr_check_list:
```

Comment:
A separate function would be good in this file only.

Reply:
Changed this code to have a separate function for boolean expressions.